File: packages/plugin-sql/src/schema/participant.ts
import { sql } from 'drizzle-orm';
import { foreignKey, index, pgTable, text, timestamp, uuid } from 'drizzle-orm/pg-core';
import { agentTable } from './agent';
import { entityTable } from './entity';
import { roomTable } from './room';

/**
 * Defines the schema for the "participants" table in the database.
 *
 * @type {import('knex').TableBuilder}
 */
export const participantTable = pgTable(
  'participants',
  {
    id: uuid('id')
      .notNull()
      .primaryKey()
      .default(sql`gen_random_uuid()`),
    createdAt: timestamp('created_at', { withTimezone: true })
      .default(sql`now()`)
      .notNull(),
    entityId: uuid('entityId').references(() => entityTable.id, {
      onDelete: 'cascade',
    }),
    roomId: uuid('roomId').references(() => roomTable.id, {
      onDelete: 'cascade',
    }),
    agentId: uuid('agentId').references(() => agentTable.id, {
      onDelete: 'cascade',
    }),
    roomState: text('roomState'),
    lastRead: timestamp('last_read', { withTimezone: true }), // For unread message tracking
  },
  (table) => [
    // Indexes for efficient queries
    index('idx_participants_entity').on(table.entityId),
    index('idx_participants_room').on(table.roomId),
    index('idx_participants_agent').on(table.agentId),
    // Foreign key constraints
    foreignKey({
      name: 'fk_room',
      columns: [table.roomId],
      foreignColumns: [roomTable.id],
    }).onDelete('cascade'),
    foreignKey({
      name: 'fk_entity',
      columns: [table.entityId],
      foreignColumns: [entityTable.id],
    }).onDelete('cascade'),
  ]
);

File: packages/plugin-sql/rust/src/schema/participant.rs
//! Participant schema for elizaOS database

/// SQL for creating the participants table
pub const CREATE_PARTICIPANTS_TABLE: &str = r#"
CREATE TABLE IF NOT EXISTS participants (
    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
    entity_id UUID REFERENCES entities(id) ON DELETE CASCADE NOT NULL,
    room_id UUID REFERENCES rooms(id) ON DELETE CASCADE NOT NULL,
    created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
    user_state TEXT
)
"#;

/// SQL for creating indexes on participants table
pub const CREATE_PARTICIPANTS_INDEXES: &str = r#"
CREATE INDEX IF NOT EXISTS idx_participants_entity_id ON participants (entity_id);
CREATE INDEX IF NOT EXISTS idx_participants_room_id ON participants (room_id);
CREATE UNIQUE INDEX IF NOT EXISTS idx_participants_unique ON participants (entity_id, room_id);
"#;

/// Participant record structure
#[derive(Clone, Debug)]
pub struct ParticipantRecord {
    pub id: uuid::Uuid,
    pub entity_id: uuid::Uuid,
    pub room_id: uuid::Uuid,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub user_state: Option<String>,
}
